tools/ed64romconfig.c

```diff
@@ -0,0 +1,197 @@
+	fprintf(stderr, "\tsram1m      Game saves to 1 megabit SRAM\n");
+	fprintf(stderr, "\tflashram    Game saves to 1 megabit FlashRAM\n");
+}
+
+bool check_flag(const char * arg, const char * shortFlag, const char * longFlag)
+{
+	return !strcmp(arg, shortFlag) || !strcmp(arg, longFlag);
+}
+
+/**
+ * Corresponds to ED64 ROM Configuration Database values:
+ * @see https://github.com/krikzz/ED64/blob/master/docs/rom_config_database.md
+ */
+uint8_t parse_save_type(const char * arg)
+{
+	if(!strcmp(arg, "none"))      return SAVETYPE_DISABLED;
+	if(!strcmp(arg, "eeprom4k"))  return SAVETYPE_EEPROM4K;
+	if(!strcmp(arg, "eeprom16k")) return SAVETYPE_EEPROM16K;
+	if(!strcmp(arg, "sram256k"))  return SAVETYPE_SRAM256K;
+	if(!strcmp(arg, "sram768k"))  return SAVETYPE_SRAM768K;
+	if(!strcmp(arg, "flashram"))  return SAVETYPE_FLASHRAM;
+	if(!strcmp(arg, "sram1m"))    return SAVETYPE_SRAM1M;
+	return SAVETYPE_NOT_SET;
+}
+
+/**
+ * Combines save_type, force_rtc, region_free values for use in the header version field:
+ * @see https://github.com/krikzz/ED64/blob/master/docs/rom_config_database.md#developer-override
+ */
+uint8_t rom_configuration(uint8_t save_type, bool force_rtc, bool region_free)
+{
+	if(!force_rtc && !region_free && save_type == SAVETYPE_NOT_SET)
+	{
+		return ROMCONFIG_NOT_SET;
+	}
+	if(save_type == SAVETYPE_NOT_SET)
+	{
+		fprintf(stderr, "WARNING: RTC/Region-Free declared without save type; defaulting to 'none'\n");
+		save_type = parse_save_type("none");
+	}
+	if(force_rtc && (save_type == SAVETYPE_EEPROM4K || save_type == SAVETYPE_EEPROM16K))
+	{
+		fprintf(stderr, "WARNING: The combination of EEPROM + RTC does not work on EverDrive!\n");
+	}
+	return save_type | (force_rtc ? 1 : 0) + (region_free ? 2 : 0);
+}
+
+int main(int argc, char *argv[])
+{
+	FILE * write_file = NULL;
+	bool force_rtc = false;
+	bool region_free = false;
+	uint8_t save_type = SAVETYPE_NOT_SET;
+	uint8_t config = ROMCONFIG_NOT_SET;
+	int i = 1;
+	const char * arg;
+
+	if(argc <= 1)
+	{
+		/* No way we can have just one argument or less */
+		print_usage(argv[0]);
+		return -1;
+	}
+	while(i < argc)
+	{
+		arg = argv[i++];
+		if(check_flag(arg, "-c", "--rtc"))
+		{
+			force_rtc = true;
+			continue;
+		}
+		if(check_flag(arg, "-r", "--regionfree"))
+		{
+			region_free = true;
+			continue;
+		}
+		if(check_flag(arg, "-w", "--savetype"))
+		{
+			if(i >= argc)
+			{
+				/* Expected another argument */
+				fprintf(stderr, "ERROR: Expected an argument to savetype flag\n\n");
+				print_usage(argv[0]);
+				return -1;
+			}
+
+			save_type = parse_save_type(argv[i++]);
+
+			if(save_type == SAVETYPE_NOT_SET)
+			{
+				/* Invalid save type */
+				fprintf(stderr, "ERROR: Invalid savetype argument\n\n");
+				print_usage(argv[0]);
+				return -1;
+			}
+
+			continue;
+		}
+		if(i < argc)
+		{
+			fprintf(stderr, "ERROR: Unexpected extra arguments\n\n");
+			print_usage(argv[0]);
+			return -1;
+		}
+
+		write_file = fopen(arg, "r+b");
```

Comment:
It looks like this could get executed multiple times in a loop. Maybe it should instead be detected as a command line syntax error

Reply:
The if block directly above aborts the command if there are any more arguments that would cause the write file to be opened more than once, but I could make this more obvious